README.md
# pypicolcd
<https://github.com/poikilos/pypicolcd>


Draw to picoLCD 256x64 and 20x4 using only pyusb (no driver required!) by importing the PicoLCD class.
![kitten](https://github.com/poikilos/pypicolcd/raw/master/screenshot.jpg)

## Main Features
* Easy: see <https://github.com/poikilos/pypicolcd/blob/master/example.py>
* Draw without driver
* Fast: refresh only refreshes zones invalidated
* Draw without dependencies other than pyusb and PIL
* Fault-tolerant: draw anything beyond range of LCD and will not crash nor miss good parts
* Image Dithering: draw color image, and it will automatically be dithered to 1-bit by luminosity (threshold is also possible)
* Persistent Framebuffer: The lcd-fb script (which runs lcdframebuffer.py) runs a server. The lcd-cli script (which runs command_line.py) is the client.
  - By default, they only communicate with localhost. To use the server over the LAN:
    - On the client, specify `lcd-cli --host ...` where `...` is the server's IP address after doing the following steps:
    - On the server, ensure incoming traffic on port 25664 is allowed.
    - On the server, specify your IP address when running the server (change the service file if you run it as a service):
```bash
lcd-fb --localhost=`hostname -i`
```
where `\`hostname -i\`` is your IP address.

## Compare
- This module does not require a (non-Python kernel) driver unlike pyusblcd in the PyPI repository.
- This module has a persistent framebuffer server unlike other driverless modules named similarly such as [JamesTheAwesomeDude/pyPicoLCD](https://github.com/JamesTheAwesomeDude/pyPicoLCD) and [itszero/picoLCD256x64](https://github.com/itszero/picoLCD256x64).
- This module uses pyusb for driverless access and implements a lcd-cli command similarly to how [rebeccaskinner/lcddeamon](https://github.com/rebeccaskinner/lcddeamon) uses libusb for driverless access and implements a usblcd command, but that daemon is written in c.
- lcd4linux is a tool that writes data using shell expressions in lcd4linux.conf, but has some features such as bold text and progress bars (which are usually used as meters).

## Install
- Uninstall lcd4linux if present: pypicolcd is not designed to work while lcd4linux is installed.
- pypicolcd uses Python 3, and though has some Python 2 considerations, is not thoroughly tested on Python 2. Therefore, make sure virtualenv is Python 3 by default, otherwise follow a guide to use the Python 3 virtualenv (such as [Installing and using virtualenv with Python 3](https://help.dreamhost.com/hc/en-us/articles/115000695551-Installing-and-using-virtualenv-with-Python-3)).
  - The client side is working with Python 2 as part of developing <https://github.com/poikilos/OctoPrint-picoLCD-Progress>.
- Install via pip to get all dependencies:
```bash
customDie() {
    echo
    echo
    echo "ERROR:"
    echo "$1"
    echo
    echo
    exit 1
}
cd ~
# if [ ! -d "lcd" ]; then
#     mkdir lcd || customDie "'mkdir lcd' failed in `pwd` as $USER"
# fi
# cd lcd || customDie "'cd lcd' failed in `pwd`."
if [ -d "lcd" ]; then
    customDie "`pwd`/lcd already exists. Try removing it, or run the remaining commands (after and not including 'virtualenv lcd') if it is already a virtualenv"
fi
virtualenv lcd || customDie "'virtualenv lcd' failed in `pwd` as $USER"
source lcd/bin/activate || customDie "'source lcd/bin/activate' failed in `pwd`"
pip install pip --upgrade || customDie "'pip install pip --upgrade' failed in `pwd` as $USER"
option=""
if [ ! -z "`pip list | grep -F pypicolcd`" ]; then
    option="--upgrade"
fi
pip install $option https://github.com/poikilos/pypicolcd/archive/master.zip || customDie "'pip install octoprint --upgrade' failed in `pwd` as $USER"
deactivate


```
- The pip install automatically gets the dependencies:
  - pyusb (formerly `sudo python3 -m pip install pyusb`)
  - PIL (formerly `sudo python3 -m pip install Pillow` or `sudo python -m pip install Pillow` or on arch, `pacman -Syu python-pillow`)

- Then you must run lcd-fb on startup. For linux, you can do the
  following:
  - `cd ~ && git clone https://github.com/poikilos/pypicolcd.git`
  - `cd pypicolcd`
  - `./install-lcd-fb.sh`
  - edit the following script if your LAN is something other
    than 192.168.1.*:
  - `"./allow port 25664 only from 192.168.1 network.sh"`
- You must also allow the unpriveleged user to send data on the USB port
  if the server prompts you with instructions on how to do that. Run
  example-cli.py or testing.pyw to view the instructions for linux.

### Update The Server
(The pip command below also updates the client.)
- First, activate the virtualenv if present such as by running `source ~/lcd/bin/activate`, then:
  `pip install --upgrade https://github.com/poikilos/pypicolcd/archive/master.zip && rm -rf ~/lcd/lib/python3.7/site-packages/pypicolcd/__pycache__/ && sudo systemctl daemon-reload && sudo systemctl restart lcd-fb`
- Or, if you have a local copy of the git repo:
  - You still must activate the venv, such as by running `source ~/lcd/bin/activate`, then do
  `pip install --upgrade ~/git/pypicolcd && rm -rf ~/lcd/lib/python3.7/site-packages/pypicolcd/__pycache__/ && sudo systemctl daemon-reload && sudo systemctl restart lcd-fb`

## Usage
* Draw Image:
  * pos is an x,y tuple
  * negative pos is allowed, which can be used for sprite animations if 64x64 cells in single-column layout or 256x64 cells in any layout
  * see comments above draw_image in picousb.py for more info
* Draw Text:
  * draw_text function's row, col format is y,x order (is considered 6x8 pixel blocks if `picolcd.dc["type"] == "graphics"`), but you can do draw_text_at to avoid that and use pixel locations directly (for text mode, these locations will be interpreted into col,row format in that order for draw_text_at function)
  * see comments above draw_text in picousb.py for more info
* Pixel manipulation:
  * For drawing many pixels at once, make your drawing faster by using `refresh_enable=False` (such as `picolcd.set_pixel(x, y, True, refresh_enable=False)`), then call `picolcd.refresh()` after all of your `set_pixel` calls are done (draw_text is an example of how to use this optimization)
  * get_pixel only works for pixels created during the life of the PicoLCD object, since it gets pixels from the offscreen pixelbuffers
* On Reconnect:
  * If you disconnect the device for 1 minute or when the minute
    changes, and the clock is on, or some other write occurs when the
    device is disconnected, the framebuffer server will know, and
    invalidate the display (so that it will redraw when possible).
    - Otherwise, you should send the `refresh` command (such as
      `lcd-cli --refresh`) to redraw the offscreen buffer.

## Known Issues
- [ ] If possible, read the state of buttons on the unit (also via pyusb).
- [ ] Add an option to disconnect from the device so it can be used by other processes.

## Troubleshooting
- Detecting resets (normally from the device being disconnected) is
  only possible if you try to write while the device is disconnected.
  Otherwise, the framebuffer (and canvas in testing.pyw) will remain
  incorrect until you call `clear()` on the PicoLCD instance (certain
  pixels involved in the command will not change if they would match the
  existing framebuffer--this is the expected behavior).

## Authors
* resources from external sources:
  * Ninepin font: by Digital Graphics Labs on <http://www.1001fonts.com/ninepin-font.html> LICENSE is "fonts/1001Fonts General Font Usage Terms.txt" except with the following specifics stated by author: Free for personal use, Free for commercial use, Modification NOT allowed, Redistribution allowed, freeware license in "fonts/ninepin/!license.txt"
  * Press Start font family (prstart.ttf, prstartk.ttf): by codeman38 on <http://www.1001fonts.com/press-start-font.html> LICENSE is "fonts/1001Fonts General Font Usage Terms.txt" except with the following specifics stated by author: Free for personal use, Free for commercial use, Modification allowed, Redistribution allowed, custom license "fonts/press-start/license.txt"
    * naturally 8px high including descenders
  * flottflott font: by Peter Wiegel on <http://www.1001fonts.com/flottflott-font.html>  LICENSE is "fonts/1001Fonts General Font Usage Terms.txt" except with the following specifics stated by author: Free for personal use, Free for commercial use, Modification allowed, Redistribution allowed, SIL Open Font License (OFL) -- see "fonts/flottflott/Open Font License.txt" and "fonts/flottflott/OFL-FAQ.txt"
  * zephyrean font family (zephyrea.ttf, zephyreg.ttf) by Ænigma Fonts on <http://www.1001fonts.com/zephyrean-font.html> LICENSE is "fonts/1001Fonts General Font Usage Terms.txt" except with the following specifics stated by author: Free for personal use, Free for commercial use, Modification NOT allowed, Redistribution allowed, custom "use them any way you want" license (see "fonts/zephyrean/zephyrean.txt")
  * ilyessuti on pixnio: CC0 licensed kitten*.jpg (2017-12-21-14-52-48) <https://pixnio.com/fauna-animals/cats-and-kittens/field-grass-cute-summer-nature-cat-outdoor-flower>, retouch poikilos -- see "images/CC0.txt"
* [poikilos](https://github.com/poikilos) code and resources not mentioned above: resources created by poikilos are CC0 -- see "images/CC0.txt")
* sphinx on excamera.com: http://excamera.com/sphinx/article-picolcd.html (61-line version of code that works with text model picoLCD 20x4 only)
  - Formerly, pypicolcd (before it was a module) required picolcd.py from http://excamera.com/sphinx/article-picolcd.html.


## Developer Notes
* Stripe caches (_s_cache) is a dict of dicts of list of bytes where each byte is a vertical stripe of 8 pixels (cached this way since device writes that way)
* To get font rect (for graphics type devices only), try something like
```python
last_rect = picolcd.draw_text(
    y, x, "|-------|",  # row,col format is in y,x order even though
                        # is pixel location if graphics type device
    erase_behind_enable=True, refresh_enable=False)
```
  then clear the LCD, or just store the numbers to a file or your code so the process doesn't need to be repeated on each run.
  Result will be a rect in format ((min_x, min_y), (max_x+1, max_y+1)) which can be later passed like `draw_text(y, x, "...", erase_rect=last_rect)`
* erase_behind_enable option will not erase beyond actual drawn rect (such as, overwriting "111" with "---" will still show the top and bottom of the 1s unless you pass erase_rect param (see above)
* contents of DC_DICT (and hence of picolcd.dc) should never be changed--they are device characteristics that define how the device operates at the lowest accessible level. If there is a different device not supported, the device should be added as a new entry in the DC_DICT, where the key is its USB device id.
* buffer state is not stored across runs, so clearing LCD should be done
  unless you only want to assign a certain slot to each program that
  displays things to it.

### What is testing.py
* Left Panel:
  * Draw Text or Draw Image (threshold will be used if Threshold is checked above the threshold value textbox)
    * If threshold unchecked, dithering will be used for Draw Image
* Right Panel:
  * Enable drawing a pixel by clicking on the canvas
  * Sends any bytes you want to your picoLCD without a driver!

### Communication Protocol
(only "picoLCD 256x64 Sideshow" was tested)

You must sent a byte array. Python will throw an exception if any integer in `this_list` is >255. `bytes(this_list)`.
In LANDSCAPE orientation:
* The term "block" here is used to refer to a 8x32 pixel area (made up of 32 bytes on the 1-bit screen buffer)
* Each byte on the chip holds 8 VERTICAL pixels
* There are 4 chips from left to right--each chip has 2 zones,
  so there are 8 zones from left to right
* There are 8 blocks (8 rows of blocks) from top to bottom (the official driver calls this a "line" which is misleading since it is an irreducible 8 pixels tall)
* long command:
  * has 11 bytes plus a length byte and data bytes
  * purpose is to select&write block&zone (but cannot select odd zones and cannot change x pixel location except in relative positioning which is not usually used with long command)
  * absolute positioning when `[8]` is 0x40 (always use 0x40 for long command)
* short command:
  * has 4 bytes plus a length byte and data bytes
  * only short command can access a block in an odd zone (right side) of a chip
  * long command must select block and chip before short command is called
  * normally, long command writes 32 bytes (0x20), then if right side of chip needs to be accessed, short command is also called
    * writes next 32 bytes (short command must be called immediately after long command writes byte 31, due to relative positioning being the only way to access the right side of the chip aka odd zone)
    * therefore, accessing a block in an odd zone always requires first filling the even zone (for example, filling zone 1 block 3 requires filling zone 0 block 3 using long command with 32 bytes then calling short command to fill zone 1 block 3, where zone 1 and block 3 are implied by relative positioning, also with 32 bytes)
* pypicolcd stores blocks separately (in picolcd.framebuffers list) so they can be sent to the graphics chip without slicing the buffer (only one framebuffer is created for text type devices)


CHANGELOG.md
 # Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a
Changelog](https://keepachangelog.com/en/1.0.0/).


## [git] - 2020-01-23
### Changed
- Rename "brightness" option to "backlight" to make room for future
  brightness option which could apply to the image brightness
  (pre-dithering multiplier).
- Move lcdclient to a separate module, and simplify stats.py by using
  lcdclient instead of command_line.
- Finish the clock option (separate x and y params for clock from
  others).


## [git] - 2020-01-23
### Added
- Allow custom text along with the lcd-stats command.

### Changed
- Improve input and output.
- (install-lcd-fb.sh) Fix service file generation.
- Fix Python 2 compatibility.
- Fix lcd-stats argument processing.
- Fix handle_error
- Fix handle_signal (close the server when OS sends a signal).
- Improve quality.sh and make PEP8 changes.
- Fix use of clear, x, and y with the lcd-stats command.
- Improve documentation including docstrings.
- Make stats have a fixed width to ensure they overwrite previous
  values.


## [git] - 2020-01-21
### Added
- Create a server so that the buffer remains the same for multiple
  programs and `lcd-cli` calls.
- Set the following new params using lcd-cli (or by sending JSON to the
  http framebuffer server directly, or by utilizing the PicoLCD
  push_action method some other way):
  - `font`
  - `x`
  - `y`
- Install as service on GNU+Linux systems using the new
  "install-lcd-fb.sh" script.

### Changed
- Clean up Draw objects for fonts (resolves issue #1 core dump in
  `draw_text` presumably).
  - This may introduce some latency, but only the first time a
    character is used during the entire run of the framebuffer server.
- `draw_lines` now accepts pixel locations as a top left corner.
- The `push_text` method (such as utilized via the `push` option) now
  acts upon `\t`, `\n`, and `\r` control characters (escaped or
  literal).
- Calculate the width of the space character (as `space_w`)
  automatically in push_text (formerly it used `3` literally).
- rename lcd-daemon to lcd-fb


## [git] - 2020-01-20
### Added
- Create a framebuffer server with a persistent connection.


## [git] - 2020-01-19
### Added
- Add a udev rules file that matches the one the program generates for a
  picoLCD 256x64 Sideshow.

### Changed
- Improve udev rule instructions
  - ATTRS, which checks parent devices, also works. ATTR does not check
    parent devices, but still works.
    - See
      <https://github.com/JamesTheAwesomeDude/pyPicoLCD/blob/full-reimplimentation/60-lcd.rules>
- Handle diconnects (see README.md).
- Fix backlight level method (and rename to `set_backlight`).
- Fix flash method.
- Allow sending bytestrings of arbitrary length using testing.pyw.


## [git] - 2020-01-17
### Added
- The pip install now includes the necessary data files (fonts and
  images) via the added MANIFEST.in for setuptools.

### Changed
- The command is now lcd-cli (It still uses command_line.py).
- The pip install now works properly, and imports work properly.
- The module name is now consistent (pypicolcd).
- Old changes from the README.md are now in this file.


## [git] - 2020-01-15
### Added
- There is now a pypicolcd-cli command (that uses command_line.py).
- Create a Changelog.

### Changed
- Refactor the project to make it into a setuptools package.
  - Adjust install instructions to match (easier and using virtualenv).
- Change case of classes from PicoLcd to PicoLCD


## [git] - 2018-01-26
### Added
- make example.py [later renamed to example-cli.ty]
- draw text
- draw image with threshold OR dithering

### Changed
- (fix use of refresh_enable) make force_refresh_enable=True work where
  on=False


pypicolcd/stats.py
#!/usr/bin/env python

import os
import sys
import platform
import copy
from datetime import datetime

from pypicolcd import lcdclient
from pypicolcd.lcdframebuffer import get_commands
from pypicolcd.lcdframebuffer import get_bool_options
from pypicolcd import to_bool

_MAX_LINES = 4


class FreeMemLinux(object):
    """
    This is a non-cross platform way to get free memory on GNU+Linux
    systems. Note that this code uses the `with ... as`, which is
    conditionally Python 2.5 compatible! If for some reason you still
    have Python 2.5 on your system add in the head of your code, before
    all imports: from __future__ import with_statement

    according to [Oz123's answer](https://stackoverflow.com/a/24186720)
    Jun 12, 2014 (edited Oct 7, 2016; accessed Jan 14, 2020)
    on <https://stackoverflow.com/questions/17718449/
    determine-free-ram-in-python>
    """

    def __init__(self, unit='kB'):

        with open('/proc/meminfo', 'r') as mem:
            lines = mem.readlines()

        self._tot = int(lines[0].split()[1])
        self._free = int(lines[1].split()[1])
        self._buff = int(lines[2].split()[1])
        self._cached = int(lines[3].split()[1])
        self._shared = int(lines[20].split()[1])
        self._swapt = int(lines[14].split()[1])
        self._swapf = int(lines[15].split()[1])
        self._swapu = self._swapt - self._swapf

        self.unit = unit
        self._convert = self._factor()

    def _factor(self):
        """determine the convertion factor"""
        if self.unit == 'bytes':
            return 1
        if self.unit.lower() == 'kb':
            return 1024
        if self.unit.lower() == 'k':
            return 1024.0
        if self.unit.lower() == 'mb':
            return 1/1024.0
        if self.unit.lower() == 'm':
            return 1/1024.0
        if self.unit.lower() == 'gb':
            return 1/1024.0/1024.0
        if self.unit.lower() == 'g':
            return 1/1024.0/1024.0
        if self.unit == '%':
            return 1.0/self._tot
        else:
            raise Exception("Unit not understood")

    @property
    def total(self):
        return self._convert * self._tot

    @property
    def used(self):
        return self._convert * (self._tot - self._free)

    @property
    def used_real(self):
        """memory used which is not cache or buffers"""
        return self._convert * (self._tot - self._free -
                                self._buff - self._cached)

    @property
    def shared(self):
        return self._convert * (self._tot - self._free)

    @property
    def buffers(self):
        return self._convert * (self._buff)

    @property
    def cached(self):
        return self._convert * self._cached

    @property
    def user_free(self):
        """This is the free memory available for the user"""
        return self._convert * (self._free + self._buff + self._cached)

    @property
    def swap(self):
        return self._convert * self._swapt

    @property
    def swap_free(self):
        return self._convert * self._swapf

    @property
    def swap_used(self):
        return self._convert * self._swapu


def show_lines_for_headless(lines):
    """
    Show lines on tty1. This writes text lines to picoLCD screens if the
    server has no video card (tested on Debian 10 netinstall with no
    special picoLCD-related packages).
    """
    # See <https://stackoverflow.com/questions/20894969/
    # python-reading-and-writing-to-tty>
    # (uses mode os.RDRW)
    tty = os.open("/dev/tty1", os.O_WRONLY)
    count = 0
    if lines is not None:
        for line in lines:
            if count < _MAX_LINES:
                if line is not None:
                    os.write(tty, str.encode("\n" + line))
                else:
                    os.write(tty, b"\n")
                count += 1
    for i in range(_MAX_LINES - count):
        os.write(tty, b"\n")
    os.close(tty)


def run_cli(args):
    # TODO: This is not tested. Use run from from pypicolcd.command_line
    # directly instead.
    results = {"error": "contacting the server did not complete"}
    # See <https://www.saltycrane.com/blog/2008/09/
    # how-get-stdout-and-stderr-using-python-subprocess-module/>
    if len(args) >= 1:
        if not args[0].endswith('lcd-cli'):
            args = ['lcd-cli'] + args
    if len(args) < 2:
        raise ValueError("You didn't provide any arguments.")
    wgproc = subprocess.Popen(args, stdout=subprocess.PIPE,
                              stderr=subprocess.STDOUT)
    (standardout, junk) = wgproc.communicate()
    result = standardout
    return result


k_div = 1024
m_div = k_div * 1024
g_giv = m_div * 1024


def freeSpaceAt(path, unit="bytes", places=2):
    """
    Get the current free space on the drive that contains a specific
    path.

    Sequential arguments:
    path -- a path under a drive, or the exact path of a drive.

    Keyword arguments:
    unit -- (default: "bytes") The unit to use: bytes, kb, mb, gb (or
        the respectively synonymous b, k, m, or g).
    places -- (default: 2) The maximum number of decimal places to
        return. It it always 0 if using bytes.

    Returns:
        Get a float (integer if bytes) representing free space.
    """
    # See Mechanical snail's Sep 8, 2012 answer (edited by Mark Amery
    # Apr 13, 2019)
    # on <https://stackoverflow.com/questions/4260116/
    # find-size-and-free-space-of-the-filesystem-containing-a-given-
    # file>
    statvfs = os.statvfs(path)
    # statvfs.f_frsize * statvfs.f_blocks     # Size in bytes
    # statvfs.f_frsize * statvfs.f_bfree      # Actual free bytes
    # statvfs.f_frsize * statvfs.f_bavail     # free bytes for users
    #                                           (excl. reserved space)
    try:
        bfree = long(statvfs.f_bfree) * long(statvfs.f_bsize)
    except NameError:
        # Python 3 is always long
        bfree = statvfs.f_bfree * statvfs.f_bsize
    ret = None
    # fmt = "0:." + str(places) + "f"
    if (unit.lower() == "bytes") or (unit.lower() == "b"):
        ret = bfree
    elif (unit.lower() == "kb") or (unit.lower() == "k"):
        ret = round(bfree/k_div, places)
    elif (unit.lower() == "mb") or (unit.lower() == "m"):
        ret = round(bfree/m_div, places)
    elif (unit.lower() == "gb") or (unit.lower() == "g"):
        ret = round(bfree/g_div, places)
    return ret


def freeSpaceAtFmt(path, unit="bytes", places=2):
    fmt = "{:." + str(places) + "f}"
    if unit == "bytes":
        fmt = "{:.0f}"
    n = freeSpaceAt(path, unit=unit)
    return fmt.format(n)


def generate_action(action, lines, x=None, y=None):
    action = copy.deepcopy(action)
    if x is not None:
        action["x"] = int(x)
    if y is not None:
        action["y"] = int(y)
    action["lines"] = lines
    return action





# def push_action_and_stats(params):

def run_and_add_stats(args):
    """
    Process args and send them along as an action dictionary. Generate
    and send additional dictionaries to show system stats on the screen
    (displayed along with any custom text that may be in args). Any arg
    that applies to the entire operation may be either used only on the
    first action or appended to every action, as appropriate. You
    should send only the slice sys.argv[1:] if using sys.argv, as even
    the first index of args will be processed, but sys.argv[0] is the
    system command (which this function will not understand).
    """

    # show_lines_for_headless(["Hello World!"])
    slash_name = "primary /"
    drive2_name = "Home"
    stat_order = ["Memory", drive2_name, slash_name]
    stats = {}

    f = FreeMemLinux()
    # print(f.total, f.used,  f.user_free)
    # f_mb = FreeMemLinux(unit='MB')
    # f_percent = FreeMemLinux(unit='%')
    paths = {}
    paths[drive2_name] = os.environ.get("HOME")
    if paths[drive2_name] is None:
        paths[drive2_name] = os.environ.get("USERPROFILE")
    # freemem = FreeMemLinux(unit='%')
    unit = 'mb'
    freemem = FreeMemLinux(unit=unit)
    stats["Memory"] = "{:.2f} {}".format(
        freemem.total - freemem.used,
        unit
    )

    if platform.system() == "Windows":
        paths[slash_name] = "C:\\"
        del paths[drive2_name]
        stat_order.remove(drive2_name)
        if os.path.isdir("D:\\"):
            drive2_name = "D:"
            paths[drive2_name] = "D:\\"
            stat_order.append(drive2_name)
        else:
            drive2_name = None

    if drive2_name is not None:
        if paths[drive2_name] is not None:
            stats[drive2_name] = "{} {}".format(
                freeSpaceAtFmt(paths[drive2_name], unit=unit),
                unit
            )
        else:
            stats[drive2_name] = "?"
    paths[slash_name] = "/"
    stats[slash_name] = "{} {}".format(
        freeSpaceAtFmt(paths[slash_name], unit=unit),
        unit
    )
    if drive2_name is not None:
        if stats[slash_name] == stats[drive2_name]:
            del stats[drive2_name]
            stat_order.remove(drive2_name)
    pfsm = 1.5  # proportionally-spaced font space multiplier
    stat_list = []
    name_max = 0
    stat_max = 0
    for name in stat_order:
        if len(name) > name_max:
            name_max = len(name)
        if len(stats[name]) > stat_max:
            stat_max = len(stats[name])
    if stat_max < 13:
        stat_max = 13
        # such as "1000000.00 mb"
        # (which is1 tb)

    name_max = int(round(float(name_max) * pfsm))
    fmt = "{:<" + str(name_max) + "} {:>" + str(stat_max) + "}"
    for name in stat_order:
        if name in stats.keys():
            stat_list.append(
                fmt.format(name + ":", stats[name])
            )
        else:
            stat_list.append(fmt.format(name, "missing"))
    params = {}

    # Parse the params for filtering only
    # --still pass them as terminal params.
    custom_lines = []
    for i in range(0, len(args)):
        arg = args[i]
        if arg.startswith("--") and not arg.startswith("---"):
            ender = len(arg)
            sign_i = arg.find("=")
            value = True
            if sign_i > -1:
                ender = sign_i
                value = arg[sign_i+1:]
            name = arg[2:ender]
            if name == "headless":
                # TODO: implement this local command on the server
                params[name] = True
            elif name in get_commands():
                params[name] = True
            elif name in get_bool_options():
                params[name] = to_bool(value)
            else:
                params[name] = value
            print("* {}={}".format(name, value))
        else:
            custom_lines.append(arg)
    x = params.get("x")
    y = params.get("y")
    clock_d = None
    batch = {}
    batch_order = ["timestamp", "stats"]
    if params.get("clock") is not None:
        # Generate clock BEFORE x and y are set forcibly below.
        clock_d = generate_action(params, None)
        batch["clock"] = clock_d
        del params["clock"]  # don't keep putting clock everywhere below
        batch_order.append("clock")
    if x is None:
        x = 0
    if y is None:
        y = 0
    top = 63 - (len(stat_order) * 8)
    stat_d = generate_action(params, stat_list, x=0, y=top)

    now = datetime.now()
    now_s = now.strftime("%Y-%m-%d %H:%M:%S")

    batch["stats"] = stat_d
    timestamp_lines = []
    timestamp_lines.append("on {}".format(os.uname()[1]))
    timestamp_lines.append("@" + now_s)
    # x=153 puts rightmost pixel of "@____-%m-%d %H:%M:%S" at the edge
    push_down = (len(stat_list)-len(timestamp_lines))*8
    timestamp_d = generate_action(params, timestamp_lines, x=153,
                                  y=top+push_down)
    batch["timestamp"] = timestamp_d

    if len(custom_lines) > 0:
        # for k, v in params.items():
        #     if (k != "clear") and (k != "y"):
        #         print("* more lines: --{}={}".format(k, v))
        #         custom_lines.append("--{}={}".format(k, v))
        custom_d = generate_action(params, custom_lines, x=y, y=y)
        batch["custom"] = custom_d
        batch_order.append("custom")
        # print("{} custom line(s)".format(len(custom_lines)))
    else:
        print("There are no custom lines.")
    first = True
    for key in batch_order:
        print("")
        action = batch[key]
        if not first:
            try:
                del action["clear"]
            except KeyError:
                pass
            try:
                del action["refresh"]
            except KeyError:
                pass
        first = False
        results = lcdclient.send_action(action)
        if results.get("status") != "OK":
            print('* {}'.format(results))
            print("  * in response to showing {}:"
                  " '{}'".format(key, action))
            if action.get('headless') is True:
                print("* attempting to write to tty1 (which could be a"
                      " picoLCD display on a headless server)...")
                if "lines" in action:
                    show_lines_for_headless(action["lines"])
        else:
            # print("  * in response to '{}'".format(action))
            info_s = results.get("info")
            if info_s is not None:
                del results["info"]
            print('* {}'.format(results))
            if info_s is not None:
                print("\n".join(info_s.split("\\n")))


def main():
    run_and_add_stats(sys.argv[1:])


if __name__ == "__main__":
    main()
